### PR TITLE
Add optional attendee support to query

### PR DESCRIPTION
**Pull request by tttzach**
Changed query to support optional attendees and added test cases with optional attendees.
Definitely not the most optimized algorithm but tried to reuse as much code as possible!

Screenshot 2020-06-19 at 3 33 48 PM

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,22 +15,31 @@
...
-    Collection<String> proposedAttendees = request.getAttendees();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
 
-    List<TimeRange> unavailableTimeRanges = getUnavailableTimeRanges(events, proposedAttendees);
-    List<TimeRange> availableTimeRanges = getAvailableTimeRanges(unavailableTimeRanges, proposedDuration);
+    List<TimeRange> mandatoryUnavailableTimeRanges = getUnavailableTimeRanges(events, mandatoryAttendees);
+    List<TimeRange> optionalUnavailableTimeRanges = getUnavailableTimeRanges(events, optionalAttendees);
+    List<TimeRange> mandatoryAvailableTimeRanges = getAvailableTimeRanges(mandatoryUnavailableTimeRanges, proposedDuration);
+    List<TimeRange> combinedTimeRanges = Stream.concat(mandatoryUnavailableTimeRanges.stream(), optionalUnavailableTimeRanges.stream()).collect(Collectors.toList());
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:36]: Nit: variable should be `combinedUnavailableTimeRanges` for consistency/clarity.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,22 +15,31 @@
...
+    List<TimeRange> combinedAvailableTimeRanges = getAvailableTimeRanges(combinedTimeRanges, proposedDuration);
 
-    return availableTimeRanges;
+    if (combinedAvailableTimeRanges.isEmpty() && !mandatoryAttendees.isEmpty()) {
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:39]: Add a short comment in the code to explain this logic. It's not obvious from just reading this code.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 3]
```diff
@@ -270,5 +273,152 @@ public void notEnoughRoom() {
...
+
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Arrays.asList();
+
+    Assert.assertEquals(expected, actual);
+  }
+}
```
Comment (antmar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:424]: Consider adding tests where the event has required attendees, but the required attendees have no events scheduled that day. I'll let you figure out what optional-attendee scenarios these cases should cover.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (tttzach) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Agreed, thanks!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (tttzach) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:40]: I agree that it's confusing, thanks!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 3 above)
Comment (tttzach) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:473]: I believe the former is the `noConflicts()` test that was in the repository by default. 
```
@Test
public void noConflicts() {
  MeetingRequest request = new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);

  Collection<TimeRange> actual = query.query(NO_EVENTS, request);
  Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);

  Assert.assertEquals(expected, actual);
}
```
I've added some tests based on the latter and they passed but my gut feeling from your suggestion after experiencing technical interviews tells me that I am missing some test case. Let me know if my suspicions are true D:

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (tttzach) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: `RESOLVED`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (tttzach) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:40]: `RESOLVED`

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 3 above)
Comment (antmar) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:472]: These are the tests I was looking for.

Comment (tttzach) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:472]: Cool, thanks!

Comment (tttzach) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:472]: `RESOLVED`

**Pull request merged**